File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-16.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // l   l       r
        // z   x   y   z   x   y   z           {z:0 , x:1 , y:2}

        // check if val in hasmap yes then change l point to max of current or val+1

        int l,r,res;
        Map<Character,Integer> map = new HashMap<>();
        l=0;
        r=0;
        res = 0;
        while(r<s.length()){
            var val = s.charAt(r);
            if(map.containsKey(val)){
                l = Math.max(l,map.get(val)+1);
            }
            map.put(val,r);
            res = Math.max(res,r-l+1);
            r++;
        }
        return res;
        

        
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-16.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        M,N = len(s),len(t)
        if N>M:
            return 0
        dp = {}
        def dfs(i,j):
            if j == N:
                return 1
            if i == M:
                return 0
            if (i,j) in dp:
                return dp[(i,j)]
            res = dfs(i+1,j)
            if s[i] == t[j]:
                res += dfs(i+1,j+1)
            dp[(i,j)] = res
            return dp[(i,j)]
        return dfs(0,0)




      




        